core/lib/Cardano/Multisig.hs

```diff
@@ -0,0 +1,128 @@
+import qualified Data.ByteArray as BA
+
+
+-- | A 'MultisigScript' type represents multi signature script. The script embodies conditions
+-- that need to be satisfied to make it valid.
+--
+-- @since 3.0.0
+data MultisigScript =
+      RequireSignatureOf VerificationKeyHash
+    | RequireAllOf [MultisigScript]
+    | RequireAnyOf [MultisigScript]
+    | RequireMOf Word8 [MultisigScript]
+    deriving stock (Generic, Show, Eq)
+
+instance NFData MultisigScript
+
+-- | A 'VerificationKeyHash' type represents verification key hash
+-- that participate in building MultisigScript. The hash is expected to have size of
+-- 28-byte.
+--
+-- @since 3.0.0
+newtype VerificationKeyHash = VerificationKeyHash ByteString
+    deriving (Generic, Show, Eq)
+
+instance NFData VerificationKeyHash
+
+fromVerificationKey :: Shelley 'MultisigK XPub -> VerificationKeyHash
+fromVerificationKey = VerificationKeyHash . blake2b224
+
+verificationKeyHashFromBytes :: ByteString -> VerificationKeyHash
+verificationKeyHashFromBytes = VerificationKeyHash . invariantSize pubkeyHashSize
+
+toCBOR' :: MultisigScript -> CBOR.Encoding
+toCBOR' (RequireSignatureOf (VerificationKeyHash verKeyHash)) =
+    encodeMultiscriptCtr 0 2 <> CBOR.encodeBytes verKeyHash
+toCBOR' (RequireAllOf contents) =
+    encodeMultiscriptCtr 1 2 <> encodeFoldable toCBOR' contents
+toCBOR' (RequireAnyOf contents) =
+    encodeMultiscriptCtr 2 2 <> encodeFoldable toCBOR' contents
+toCBOR' (RequireMOf m contents) =
+    encodeMultiscriptCtr 3 3 <> CBOR.encodeInt (fromInteger $ toInteger m)
+    <> encodeFoldable toCBOR' contents
+
+-- | This function realizes what cardano-node's `Api.serialiseToCBOR script` realizes
+-- This is basically doing the symbolically following:
+-- toCBOR [0,multisigScript]
+toCBOR :: MultisigScript -> ByteString
+toCBOR script =
+    CBOR.toStrictByteString $ encodeMultisigBeginning <> toCBOR' script
+
+encodeMultisigBeginning :: CBOR.Encoding
+encodeMultisigBeginning = CBOR.encodeListLen 2 <> CBOR.encodeWord8 0
+
+encodeMultiscriptCtr :: Word -> Word -> CBOR.Encoding
+encodeMultiscriptCtr ctrIndex listLen =
+    CBOR.encodeListLen listLen <> CBOR.encodeWord ctrIndex
+
+encodeFoldable :: (Foldable f) => (a -> CBOR.Encoding) -> f a -> CBOR.Encoding
+encodeFoldable encode xs = wrapArray len contents
+  where
+    (len, contents) = foldl' go (0, mempty) xs
+    go (!l, !enc) next = (l + 1, enc <> encode next)
+
+    wrapArray :: Word -> CBOR.Encoding -> CBOR.Encoding
+    wrapArray len' contents' =
+        if len' <= 23
+        then CBOR.encodeListLen len' <> contents'
+        else CBOR.encodeListLenIndef <> contents' <> CBOR.encodeBreak
+
+-- | This function realizes what cardano-node's
+-- `Api.serialiseToRawBytes $ Api.scriptHash script` realizes
+-- This is basically doing the symbolically the following (using symbols from toCBOR):
+-- digest $ (nativeMultisigTag <> toCBOR multisigScript )
+toScriptHash :: MultisigScript -> ByteString
```

Comment:
There's a dedicated type defined for `VerificationKeyHash` above but we use a raw ByteString here, why :thinking: ? What about using newtypes for both and deriving `ByteArrayAccess` on both to make conversion to bytestrings easy?

Reply:
I wanted to enforce 28-byte payload in `VerificationKeyHash`...

Reply:
did something pertinent in https://github.com/input-output-hk/cardano-addresses/pull/76/commits/b6537636e737b126cd8440a522451df694129c2e